### PR TITLE
Shorten UTD marker file

### LINKED ISSUE #9346: CopyUpToDateMarker path too long
reported by @olgaark:

> We’ve recently had to change the IntDir (IntermediateOutputPath) default value for some vc projects (which made it longer by 17 chars max) and now hitting “CopyUpToDateMarker path too long” problem in some of our tests.
> 
> Bug 1900392: Unable to read the project file because the expression exceeds the OS max path limit
> 
> The item is added in MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets
> 
>   <ItemGroup>
>     <IntermediateAssembly Include="$(IntermediateOutputPath)$(TargetName)$(TargetExt)"/>
>     <FinalDocFile Include="@(DocFileItem->'$(OutDir)%(Filename)%(Extension)')"/>
>     <CopyUpToDateMarker Include="$([MSBuild]::NormalizePath('$(MSBuildProjectDirectory)', '$(IntermediateOutputPath)', '$(MSBuildProjectFile).CopyComplete'))" />
>   </ItemGroup>
> 
> So two questions:
> 1.	Can the path be shortened by using short(er) file name? $(MSBuildProjectFile) can be quite long and using it in other paths is often problematic. VC is currently using ShortProjectName in path’s defaults. Can something similar or Windows short 8.3 file name be used in msbuild common targets for paths too?
> 2.	VC projects override GetTargetPathWithTargetPlatformMoniker target where it is used in the CurrentVersion.targets’ implementation. Do we need to add it in the VC’s override?
> 

The item in question: https://github.com/dotnet/msbuild/blob/main/src/Tasks/Microsoft.Common.CurrentVersion.targets#L392

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #9346, #7131

### Context
With long project paths CopyUpToDate mareker file can get long and break build in VS as long file support is not on.
This is a proposal of short/mid term remedy by shortening the marker file path a bit. However it applies only to non-CPS cases and it applies to CopyUpToDate marker file only.
Considering those facts it's questionable if this should be part of the MSBuild repo.
Alternatively we can decide to fix it more widely by renaming the `MSBuildCopyMarkerName` to `MSBuildProjectMarkerName` and use it in various other scenarios where marker/cache files are generated in a similar way - though that would have much higher breaking possibility.

### CODE REVIEW

**Comment by JanKrivanek** [comment]
> Do you have an experimental insertion for this fix? It would be nice to see how if affects VS in advance.

https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/509283

I'm looking into the checks failures.
Did you have anything else specific in your mind to watch for?

**Comment by JanKrivanek** [comment]
Closing in order to rekick the confused git pull

**Comment by JanKrivanek** [comment]
CI blocked by ADO issue. IcM [internal]: https://portal.microsofticm.com/imp/v3/incidents/details/439851463/home

File: src/Build/Utilities/FowlerNollVo1aHash.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                    char ch = text[i];
+                    byte b = (byte)ch;
+                    hash ^= b;
+                    hash *= fnvPrimeA64Bit;
+
+                    b = (byte)(ch >> 8);
+                    hash ^= b;
+                    hash *= fnvPrimeA64Bit;
+                }
+            }
```
Comment (ladipro) [inline src/Build/Utilities/FowlerNollVo1aHash.cs:94]: nit: This can potentially have a slightly faster implementation if we care enough.

```suggestion
            ReadOnlySpan<byte> span = MemoryMarshal.Cast<char, byte>(text);
            for (int i = 0; i < span.Length; i++)
            {
                hash = unchecked((hash ^ span[i]) * fnvPrimeA64Bit);
            }
```

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 2]
```diff
@@ -385,11 +385,19 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
     <_GenerateBindingRedirectsIntermediateAppConfig>$(IntermediateOutputPath)$(TargetFileName).config</_GenerateBindingRedirectsIntermediateAppConfig>
   </PropertyGroup>
+  
+  <PropertyGroup>
+    <MSBuildCopyMarkerName Condition ="'$(MSBuildCopyMarkerName)' == ''">$(MSBuildProjectFile)</MSBuildCopyMarkerName>
+	<!-- For a long MSBuildProjectFile let's shorten this to 17 chars - using the first 8 chars of the filename and either the ProjectGuid if it exists -->
+    <MSBuildCopyMarkerName Condition ="'$(MSBuildCopyMarkerName.Length)' &gt; '17' and '$(ProjectGuid)' != ''">$(MSBuildProjectFile.Substring(0,8)).$(ProjectGuid.Substring(1,8))</MSBuildCopyMarkerName>
+	<!-- or a filename hash if the guid is not present (in such case the filename was not shortened and is still over 17 chars long). -->
+	<MSBuildCopyMarkerName Condition ="'$(MSBuildCopyMarkerName.Length)' &gt; '17'">$(MSBuildProjectFile.Substring(0,8)).$([MSBuild]::StableStringHash($(MSBuildProjectFile)).ToString("X8"))</MSBuildCopyMarkerName>
+  </PropertyGroup>
```
Comment (ladipro) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:395]: super-nit: Indentation in this `PropertyGroup` doesn't look right.

File: src/Build/Utilities/FowlerNollVo1aHash.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.Build.Utilities
+{
+    internal static class FowlerNollVo1aHash
```
Comment (ladipro) [inline src/Build/Utilities/FowlerNollVo1aHash.cs:6]: The [StringTools library](https://github.com/dotnet/msbuild/tree/main/src/StringTools) would be a slightly preferred place for this.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 4]
```diff
@@ -385,11 +385,19 @@ Copyright (C) Microsoft Corporation. All rights reserved.
...
 
   <ItemGroup>
     <IntermediateAssembly Include="$(IntermediateOutputPath)$(TargetName)$(TargetExt)"/>
     <FinalDocFile Include="@(DocFileItem->'$(OutDir)%(Filename)%(Extension)')"/>
-    <CopyUpToDateMarker Include="$([MSBuild]::NormalizePath('$(MSBuildProjectDirectory)', '$(IntermediateOutputPath)', '$(MSBuildProjectFile).CopyComplete'))" />
+    <CopyUpToDateMarker Include="$([MSBuild]::NormalizePath('$(MSBuildProjectDirectory)', '$(IntermediateOutputPath)', '$(MSBuildCopyMarkerName).Up2Date'))" />
```
Comment (ladipro) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:400]: I believe it would be safer to append the new extension to the property and use just `$(MSBuildCopyMarkerName)` here. That way users who, for whatever reason, want to keep the old filename can actually do it.

File: src/Build/Utilities/FowlerNollVo1aHash.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Build/Utilities/FowlerNollVo1aHash.cs:94]: Love it!! Adopting..

File: src/StringTools/FowlerNollVo1aHash.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            return hash;
+        }
+
+        /// <summary>
+        /// Combines two 64 bit hashes into one.
+        /// </summary>
+        /// <param name="left"></param>
+        /// <param name="right"></param>
+        /// <returns></returns>
+        [CLSCompliant(false)]
```
Comment (ladipro) [inline src/StringTools/FowlerNollVo1aHash.cs:127]: Is this needed? Also nit: Missing doc-comments for the parameters.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets [hunk 6]
```diff
@@ -385,11 +385,20 @@ Copyright (C) Microsoft Corporation. All rights reserved.
   <PropertyGroup>
     <_GenerateBindingRedirectsIntermediateAppConfig>$(IntermediateOutputPath)$(TargetFileName).config</_GenerateBindingRedirectsIntermediateAppConfig>
   </PropertyGroup>
+  
+  <PropertyGroup Condition ="'$(MSBuildCopyMarkerName)' == ''">
```
Comment (ladipro) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:389]: nit: There's a space before but not after `=`. The convention in this file seems to be no spaces.

**Review by ladipro — APPROVED** [review]

**Review by f-alizada — APPROVED** [review]
Thank you for the explanation :)

File: src/StringTools/FowlerNollVo1aHash.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                    hash ^= b;
+                    hash *= fnvPrimeA64Bit;
+                }
+            }
+#else
+            ReadOnlySpan<byte> span = MemoryMarshal.Cast<char, byte>(text.AsSpan());
+            foreach (byte b in span)
+            {
+                hash = unchecked((hash ^ b) * fnvPrimeA64Bit);
+            }
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/StringTools/FowlerNollVo1aHash.cs:115]: @KalleOlaviNiemitalo - Thank you for bringing this up! 
We're not declaring what can be the expectations of stableness of the hashing between different systems (we actually do not declare anything ... yet :-)) - but it'd certainly be nice to have it more universal.

@uweigand - would you be willing to chime in and share your thoughts on the following?

Basically - since the `MemoryMarshal.Cast<char, byte>` changes how we fetch data from memory to processor - it'll behave differently between Endinan-ness representations. But how anout the version which manually splits each char to bytes:
```csharp
for (int i = 0; i < text.Length; i++)
{
    char ch = text[i];
    byte b = (byte)ch;
    hash ^= b;
    hash *= fnvPrimeA64Bit;

    b = (byte)(ch >> 8);
    hash ^= b;
    hash *= fnvPrimeA64Bit;
}
```
since all this happens in processor after the data are fetched from memory - it should fetch same results regardless of endinan-ness - is that right?

[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/StringTools/FowlerNollVo1aHash.cs:115]: Thank you for all the detailed input! I'll remove the Spans version.

The Spans version is indeed not superior (it has slight to no advantage for large strings, but performs worse for a very short strings). It's nice that it doesn't need unchecked code and is more brief - but given the fact that it can perform worse and that the other version anyways need to be maintained - it's an easy choice.

I'll as well fix the description for the `ComputeHash64Fast`
